tests/integration/utils.py

```diff
@@ -0,0 +1,239 @@
+            token = res.json()['access_token']
+        except KeyError:
+            sys.exit('Error: invalid OAuth2 credentials')
+        session.headers = {'Authorization': f'Bearer {token}'}
+
+    return session
+
+
+# Get response content in JSON format
+def get_json_content(self, response):
+    try:
+        return response.json()
+    except json.decoder.JSONDecodeError:
+        self.fail('Response not in JSON format')
+
+
+# Get resource schema from OpenAPI specification
+def get_resource_schema(self, resource):
+    return self.openapi['definitions'][resource]['properties']
+
+
+# Helper function to make a web request and lightly validate the response
+def make_request(self, endpoint, expected_status_code,
+                 params=None,
+                 max_elapsed_seconds=5):
+    """
+    Keyword arguments:
+    * endpoint -- the endpoint to request
+    * expected_status_code -- expected HTTP status code
+    * params -- [optional] key-value pairs parameters (default: None)
+    * max_elapsed_seconds -- [optional] maximum elapsed times (default: 5)
+    Return:
+    A response object contains a server’s response to an HTTP request
+    """
+
+    requested_url = f'{self.base_url}{endpoint}'
+    response = self.session.get(requested_url, params=params)
+    logging.debug(f'Sent request to {requested_url}')
+    # Response status code should be as expected
+    status_code = response.status_code
+    self.assertEqual(status_code, expected_status_code)
+
+    # Response time should less then max_elapsed_seconds
+    elapsed_seconds = response.elapsed.total_seconds()
+    logging.debug(f'Request took {elapsed_seconds} second(s)')
+    self.assertLess(elapsed_seconds, max_elapsed_seconds)
+
+    return response
+
+
+# Check the schema of response match OpenAPI specification
+def check_schema(self, response, schema):
+    # Helper function to get attributes of the schema
+    def __get_schema_attributes():
+        return schema['attributes']['properties']
+
+    # Helper function to map between OpenAPI data types and python data types
+    def __get_attribute_type(attribute):
+        types_dict = {
+            'string': str,
+            'integer': int,
+            'int32': int,
+            'int64': int,
+            'float': float,
+            'double': float,
+            'boolean': bool,
+            'array': list,
+            'object': dict
+        }
+        if '$ref' in attribute:
+            openapi_type = __get_object_type(attribute['$ref'], types_dict)
+        elif 'properties' in attribute:
+            return dict
+        elif 'format' in attribute and attribute['format'] in types_dict:
+            openapi_type = attribute['format']
+        elif 'type' in attribute:
+            openapi_type = attribute['type']
+
+        if openapi_type:
+            return types_dict[openapi_type]
+
+        logging.warning('OpenAPI property contains no type or properties')
+        return None
```

Comment:
Since `openapi_type` is only declared in the previous conditionals, encountering a property with no type or properties will cause a runtime error. Also, it looks like the logging warning is repeated in `__get_object_type()` so it looks like this could be refactored a bit.